LoadLibrary refactoring

swaroop-sridhar opened the pull request:
This change refactors the code in DllImport in preparation
for implementing the new  NativeLibrary API here:
dotnet/corefx#32015

In particular, it introduces a change in the semantics of the
internal LoadLibrary helper functions.

When a native library is loaded, there are two categories of callers
expecting different return values:
 * External callers like AssemblyNative::InternalLoadUnmanagedDllFromPath()
   and the upcoming System.Runtime.Interop.Marshall.LoadLibrary()
   need the raw system handle
 * Internal callers like LoadLibraryModule() need the PAL registered handle

This change modifies the internal LoadLibraryModule* methods to work
in terms of native system handles, so that external callers can obrain
them directly. Methods requiring PAL-handles can register them explicitly.

[PS: NDirect::LoadLibraryFromPath was already written to return the
system handle instead of the PAL handle. This change extends
the behavior to other private members.]

There is no change in external signature of DllImport class, or the
native Dll cache in AppDomain class.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6069,7 +6069,13 @@ class LoadLibErrorTracker
-//  Local helper function for the LoadLibraryModule function below
+// Local helper function to load a library.
+// Load the library directly, and don't register it yet with PAL. Returns the native system libray handle.
+// * External callers likeAssemblyNative::InternalLoadUnmanagedDllFromPath() and the upcoming 
```
"like AssemblyNative::InternalLoadUnmanagedDllFromPath()"

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6069,7 +6069,13 @@ class LoadLibErrorTracker
+// Load the library directly, and don't register it yet with PAL. Returns the native system libray handle.
+// * External callers likeAssemblyNative::InternalLoadUnmanagedDllFromPath() and the upcoming 
+//   System.Runtime.Interop.Marshall.LoadLibrary() need the raw system handle
+// * Internal callers like LoadLibraryModule() can obrain it via PAL_RegisterLibraryDirect()
```
"obtain"

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -6473,99 +6414,171 @@ HINSTANCE NDirect::LoadLibraryModule(NDirectMethodDesc * pMD, LoadLibErrorTracke
-            *szComma = '\0';
-            while (COMCharacter::nativeIsWhiteSpace(*(++szComma)));
+        *szComma = '\0';
+        while (COMCharacter::nativeIsWhiteSpace(*(++szComma)));
```
Please move the `;` to the next line _or_ comment that we are trimming here.

janvorli commented:
@swaroop-sridhar I am sorry for the delay, I was focusing on other stuff. I am planning to review this later today.

janvorli commented on src/vm/dllimport.cpp:
```diff
@@ -6069,7 +6069,13 @@ class LoadLibErrorTracker
+//   System.Runtime.Interop.Marshall.LoadLibrary() need the raw system handle
+// * Internal callers like LoadLibraryModule() can obrain it via PAL_RegisterLibraryDirect()
+//
+// This method returns native system libray handle not registered with the PAL
```
A nit: libray -> library (multiple places)

janvorli commented on src/vm/dllimport.cpp:
```diff
@@ -6473,99 +6414,171 @@ HINSTANCE NDirect::LoadLibraryModule(NDirectMethodDesc * pMD, LoadLibErrorTracke
+        // First checks if the method has DefaultDllImportSearchPathsAttribute. If method has the attribute
+        // then dllImportSearchPathFlag is set to its value.
+        // Otherwise checks if the assembly has the attribute. 
+        // If assembly has the attribute then flag ise set to its value.
```
A nit: ise -> is

jkotas commented:
> added comments describing when the HMODULE 

Or even better: define `NATIVE_LIBRARY_HANDLE` to make this self-documenting.

swaroop-sridhar commented:
>> Or even better: define NATIVE_LIBRARY_HANDLE to make this self-documenting. 

I thought about this, but held back because the types defined in `pal_mstypes.h` were a subset of 
https://docs.microsoft.com/en-us/windows/desktop/winprog/windows-data-types 

I'll make the change now.

swaroop-sridhar commented:
Thanks for the feedback @janvorli @jkotas @AaronRobinsonMSFT 
I've addressed the feedback in https://github.com/dotnet/coreclr/pull/20841/commits/73012b5321064af0b9975e38719c42d668f8797a

janvorli commented on src/pal/inc/pal.h:
```diff
@@ -68,6 +68,12 @@ extern "C" {
+// In Windows, NATIVE_LIBRARY_HANDLE is the same as HMODULE.
+// On Unix systems, NATIVE_LIBRARY_HANDLE type represents a library handle not registered with the PAL.
+// To get a HMODULE on Unix, call PAL_RegisterLibraryDirect() on a NATIVE_LIBRARY_HANDLE.
+typedef HANDLE NATIVE_LIBRARY_HANDLE;
```
A nit - the dlopen returns void*, so it would be cleaner to define it that way.

janvorli replied:
Also, I cannot see the type definition for Windows. The pal.h is included for Unix only.

swaroop-sridhar replied:
I added a definition in `common.h`, for the `!FEATURE_PAL` case Thanks

jkotas replied:
inc/palclr.h would be a better place for it. It is where all similar definitions are.

swaroop-sridhar replied:
Thanks, I changed the Windows definition to be in palclr.h via palclr_win.h

janvorli commented on out:
```diff
@@ -0,0 +1,2035 @@
+Commencing CoreCLR Repo build
```
Please remove this file that was accidentally added.

swaroop-sridhar commented:
@dotnet-bot  test dotnet-coreclr

swaroop-sridhar commented:
@dotnet-bot  test Ubuntu x64 Checked CoreFX Tests

swaroop-sridhar commented:
@dotnet-bot test this please
